modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -295,139 +319,252 @@ public void finishFull(HybridTimestampTracker timestampTracker, UUID txId, boole
-                    commit ? verifyCommitTimestamp(enlistedGroups, commitTimestamp) : completedFuture(null);
-
-            return verificationFuture.handle(
-                    (unused, throwable) -> {
-                        Collection<ReplicationGroupId> replicationGroupIds = new HashSet<>(enlistedGroups.keySet());
-
-                        boolean verifiedCommit = throwable == null && commit;
-
-                        TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
-                                .txId(txId)
-                                .timestampLong(clock.nowLong())
-                                .groupId(commitPartition)
-                                .groups(replicationGroupIds)
-                                // In case of verification future failure transaction will be rolled back.
-                                .commit(verifiedCommit)
-                                .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
-                                .term(term)
-                                .build();
-
-                        return replicaService.invoke(recipientNode, req).thenRun(
-                                () -> {
-                                    updateTxMeta(txId, old -> {
-                                        if (isFinalState(old.txState())) {
-                                            finishingStateMeta.txFinishFuture().complete(old);
-
-                                            return old;
-                                        }
-
-                                        assert old instanceof TxStateMetaFinishing;
-
-                                        TxStateMeta finalTxStateMeta = coordinatorFinalTxStateMeta(verifiedCommit, commitTimestamp);
-
-                                        finishingStateMeta.txFinishFuture().complete(finalTxStateMeta);
-
-                                        return finalTxStateMeta;
-                                    });
-
-                                    if (verifiedCommit) {
-                                        observableTimestampTracker.update(commitTimestamp);
-                                    }
-                                });
-                    })
-                    .thenCompose(Function.identity())
-                    // verification future is added in order to share proper exception with the client
-                    .thenCompose(r -> verificationFuture);
-        };
-
-        AtomicReference<CompletableFuture<Void>> ref = new AtomicReference<>();
+        AtomicBoolean performingFinish = new AtomicBoolean();
         TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
             if (tuple0 == null) {
                 tuple0 = new TxContext(); // No writes enlisted.
             }
 
-            if (tuple0.finishFut == null) {
-                tuple0.finishFut = new CompletableFuture<>();
-                ref.set(tuple0.finishFut);
+            if (!tuple0.isTxFinishing()) {
```

Comment:
Currently, it is impossible to double-finish one transaction. I would replace the `if` clause with `assert !tuple0.isTxFinishing()`.